### ISSUE TITLE
Communication with compiler

### ISSUE THREAD

**Issue by natew** [steelbrain/steel-flint#21]
Ok so we want to do a variety of things:

1- Smart syntax highlighting

In the long run I actually would love to have a few things here:
- When using `prop` declarator highlight props in different color from variables

Obviously highlight views well

2- Linting

We'd want basic linting eventually but this is lower priority. When we plug in Flow that will catch a lot of lint type errors, leaving basically white space stuff which is good but not super high priority.

3- Autocomplete

Styles with css. Eventually replace emmet style tag completions because it really isn't great with Flint.

4- Refactors

Think: rename view and it renames everywhere.

5- Jump
- Jump between styles/tag
- Jump to view

6- Show in editor

From the browser right click and it jumps to view in editor

7- Intelligence

Show information about imported stuff or inline stuff by hovering. Example is function arguments.

**Comment by steelbrain** [comment]
There are number of ways to achieve this, each has it's pros and cons. I have two of them in mind right now

1 - The Push method. This approach is pretty straight forward
- The daemon knows all the views in all files
- The daemon only notifies the client of the position of views, the client figures out the rest (well, most of it) by itself
- The daemon triggers `view:added` and `view:removed` events, on the added event it emits a unique id, a `starting` and `closing` position. This way the client can add a marker to that range for further calculations. The client does not need to be notified of the change within that view, as it'll have the marker that'll constantly update itself along with the changes. The marker will be disposed on `view:removed` event which will contain the same unique ID from early on
- If the connection between the client and daemon is disrupted or the registry of views is lost on client side, the client can request is again and it'll be given to it on `view:list` event.

Knowing the boundaries of views has many plus points, we can
- Extract name of view easily when it changes because we have a starting position, we don't have to depend on the compiler for name changes
- Enable checks that provide advance intentions
- Try to figure out the basic calculations on our own without bugging the server
- Guess roles or types of basic things like JSX
- Respond to user's request a bit quickly than the other approach

2 - The pull method. This approach makes the client side implementation simpler
- The daemon does not notify the client on any view creation/deletion event
- On each intentions trigger, we have to pull the info from daemon if it's a view and if it is, where do we stand
- Things we figure out using this way are more accurate (and probably easy?)
- We have to ask daemon if we're in a view on each autocomplete call (we could theoretically cache the results, but still)

This method has a downside that it's a tiny bit slower, but has up sides like
- It's much more easier to implement on client side
- It's less error prone

FlowJS and HH both use the pull method, I am neutral about them both, which would you prefer?

**Comment by natew** [comment]
For the push method. So you would track the changes that happen to start/end within atom. We would only send view:add and remove and their position?

We'd eventually add more things like jump between tags/styles, and sending over dead styles start/end pos. That stuff seems like it would require push from compiler as well, though doesn't interfere with this.

Pull seems less powerful overall, rather than thinking in data structures we're thinking in terms of messages (we have to design custom messages for a bunch of stuff), whereas we could just send a big data structure file over per-view or per-file.

Besides potential speed issues, it seems push would be just as accurate. We'd just need to send over the right data.

Thoughts?

For reference, this is the current meta information we are tracking for previous plugin:

```
let meta = {
  [viewName]: {
    data: { file: name : str },
    styles: { [name]: locEnd: int },
    els: { [name + key]: locEnd : int }
  }
}
```

**Comment by natew** [comment]
Following up: We could send over some sort of data structure that lets you query positions easily. Not an expert in this case but it seems almost all of our needs center around figuring out what is where.

**Comment by steelbrain** [comment]
@natew Yes for the push method, all we need is the starting and ending position of a view, for example, if we have a file like this

``` js
1 > 'use strict'
2 > view Main {
3 >   <div>Something</div>
4 > }
```

it would emit a `view:added` event and would send metadata for it like

``` js
{
  starts: [Zero Indexed Row, Zero Indexed Column],
  ends: [Zero Indexed Row, Zero Indexed Column],
  uniqueId: 'some-unique-id'
}
```

That's all we need from the server to start tracking views in the push method

**Comment by natew** [comment]
```
let data = [
  { starts: [], ends: [], type: 'view', data: {} },
]
```

How does that look? The array would be sorted by `starts[0]` key. That way we could have a standard way to find anything you want given your current cursor position (probably could be improved, thoughts?)

Edited: same structure as yours, just in an array and with data/type key. What does unique ID do? Is it consistent across compiles?

**Comment by steelbrain** [comment]
@natew That would work, but when the package is going to use it, it's not going to use it like that. It is instead going to create a buffer marker with the given positions, the benefit would be that the buffer marker's positions would be the latest on buffer, we won't have to re-query the server for new coords on each change.

Therefore I prefer it the way we're going to feed it to `Range` constructor, range constructor accepts two points, `new Range([[start range, start column], [end row, end column]])`, row and column in first and row in second is 0-based, the end column however is 1-based

**Comment by natew** [comment]
Ah of course Atom has stuff for this already. Ok, I'll get you those positions. They already come through `editor:element`, with key `view` but it only has style and element information.

I will patch to add the actually view information as well.

**Comment by natew** [comment]
So I've gotten wrapped up in tons of optimization stuff that we had broken, and then fixed a what seemed like small correctness issue that ballooned to being a very large thing.

That said, along the way I updated our meta sending.

`file:meta` sends a nicely formatted object each time a file comes through, with views.

I know this isn't what you've requested. Next step is for me to build a small system that takes that out. I unfortunately have a few more things on my side to get demo ready.

Perhaps you'd be interested in jumping into flint code? This would be a good one I think (I'll attempt to get it done tm morning if not).

Basically the `onMeta.js` file inside the flint repo is what you want. The function at the bottom get's all the meta. We probably want a new file next to it `viewMessages` or something like that, `onMeta` would pass in the meta to it. It would then have a simple cache of file => views, and do the logic, then send a `bridge.message` appropriately. All of these things should be easily searchable!

**Comment by steelbrain** [comment]
Outdated since v2.0.0
